datastore/threads/threadproxyds.nim

```diff
@@ -31,15 +31,17 @@ type
   ErrorEnum {.pure.} = enum
     DatastoreErr, DatastoreKeyNotFoundErr, CatchableErr
 
-  ThreadTypes = void | bool | SomeInteger | DataBuffer | tuple | Atomic
-  ThreadResult[T: ThreadTypes] = Result[T, DataBuffer]
-
-  TaskCtx[T: ThreadTypes] = object
-    ds: ptr Datastore
-    res: ptr ThreadResult[T]
-    cancelled: bool
+  ThreadTypes* = void | bool | SomeInteger | DataBuffer | tuple | Atomic
+  ThreadResult*[T: ThreadTypes] = Result[T, DataBuffer]
+
+  TaskCtx*[T: ThreadTypes] = object
+    head: int
+    ds*: ptr Datastore
+    res*: ThreadResult[T]
+    cancelled: Atomic[bool]
+    isActive: Atomic[bool]
     semaphore: AsyncSemaphore
```

Comment:
Why would you want to do this?

Reply:
To keep the result and other args as valid memory in the event that a cancellation occurs and the thread-task is using the memory. To do that you need to ensure the lifetime of the `TaskCtx` is for as long as the anyone who is using this object. 

As I mentioned in my comment, instead of turning `TaskCtx` into a ref, you could keep it in the Future's. In that event to do it safely you'd need to do a spinlock or similar on the `isActive` until the task-thread can be cancelled or finished.